File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/messages/ProofRequestMessage.java
package msdk.java.messages;

import com.evernym.sdk.vcx.VcxException;
import com.evernym.sdk.vcx.utils.UtilsApi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.concurrent.ExecutionException;

import msdk.java.utils.CommonUtils;

public class ProofRequestMessage {
    public String threadId;
    public String name;
    public JSONObject attributes;
    public String proofReq;

    public ProofRequestMessage(String threadId, String name, JSONObject attributes, String proofReq) {
        this.threadId = threadId;
        this.name = name;
        this.attributes = attributes;
        this.proofReq = proofReq;
    }

    public static ProofRequestMessage parse(String message) {
        try {
            JSONObject json = new JSONObject(message);
            String threadId = CommonUtils.getThreadId(json);
            String name = json.getString("comment");
            JSONObject attachment = decodeProofRequestAttach(json);
            JSONObject requestedAttrs = ProofRequestMessage.extractRequestedAttributesFromProofRequest(attachment);
            return new ProofRequestMessage(threadId, name, requestedAttrs, message);
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static JSONObject decodeProofRequestAttach(JSONObject proofAttach) {
        try {
            String attachment = UtilsApi.vcxExtractAttachedMessage(proofAttach.toString()).get();
            return CommonUtils.convertToJSONObject(attachment);
        } catch (VcxException | InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static JSONObject extractRequestedAttributesFromProofRequest(JSONObject proofRequest) {
        try {
            if (proofRequest != null) {
                return proofRequest.getJSONObject("requested_attributes");
            }
            return null;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String extractRequestedNameFromProofRequest(JSONObject decodedProofAttach) {
        try {
            if (decodedProofAttach != null) {
                return decodedProofAttach.getString("name");
            }
            return null;
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/messages/ProofRequestMessage.kt
package msdk.kotlin.sample.messages

import com.evernym.sdk.vcx.utils.UtilsApi
import msdk.kotlin.sample.utils.CommonUtils
import org.json.JSONException
import org.json.JSONObject


class ProofRequestMessage(
    var threadId: String,
    var name: String,
    var attributes: JSONObject,
    var proofReq: String?
) {

    companion object {
        fun parse(message: String?): ProofRequestMessage? {
            return try {
                val json = JSONObject(message)
                val threadId: String = CommonUtils.getThreadId(json)!!
                val name = json.getString("comment")
                val attachment = decodeProofRequestAttach(json)
                val requestedAttrs =
                    extractRequestedAttributesFromProofRequest(attachment)
                ProofRequestMessage(threadId, name, requestedAttrs!!, message)
            } catch (e: JSONException) {
                e.printStackTrace()
                null
            }
        }

        fun decodeProofRequestAttach(proofAttach: JSONObject): JSONObject? {
            try {
                val attachment = UtilsApi.vcxExtractAttachedMessage(proofAttach.toString()).get()
                return CommonUtils.convertToJSONObject(attachment)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return null
        }

        fun extractRequestedAttributesFromProofRequest(proofRequest: JSONObject?): JSONObject? {
            try {
                return proofRequest?.getJSONObject("requested_attributes")
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return null
        }

        fun extractRequestedNameFromProofRequest(decodedProofAttach: JSONObject?): String? {
            try {
                return decodedProofAttach?.getString("name")
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return null
        }
    }

}